pkg/provisioningv2/rke2/planner/encryptionkeyrotation.go

```diff
@@ -96,192 +136,309 @@ func shouldRestartEncryptionKeyRotation(cp *rkev1.RKEControlPlane) bool {
+			// secrets-encrypt command was run on another node, restart
+			logrus.Infof("Leader [%s] with %s stage and follower [%s] with %s stage", leader.Machine.Status.NodeRef.Name, leaderStage, entry.Machine.Status.NodeRef.Name, stage)
+			return p.restartEncryptionKeyRotation(cp)
 		}
 	}
 
-	if cp.Status.RotateEncryptionKeysGeneration == cp.Spec.RotateEncryptionKeys.Generation &&
-		leaderPhase == EncryptionKeyRotationReencryptFinished {
-		cp.Status.RotateEncryptionKeysGeneration = cp.Spec.RotateEncryptionKeys.Generation
-		cp.Status.RotateEncryptionKeysPhase = rkev1.RotateEncryptionKeysPhaseDone
-		_, err = p.rkeControlPlanes.UpdateStatus(cp)
-		if err != nil {
-			return err
-		}
-		return generic.ErrSkip
+	// a phase has been completed at least once for this generation
+	if leaderStage == encryptionKeyRotationStageReencryptFinished &&
+		cp.Status.RotateEncryptionKeysStage != "" {
+		return p.encryptionKeyRotationFinished(cp)
 	}
 
-	if cp.Status.RotateEncryptionKeysGeneration != cp.Spec.RotateEncryptionKeys.Generation {
-		cp.Status.RotateEncryptionKeysGeneration = cp.Spec.RotateEncryptionKeys.Generation
-	}
+	cp.Status.RotateEncryptionKeysStage = leaderStage
 	err = p.encryptionKeyRotationUpdateControlPlanePhase(cp, rkev1.RotateEncryptionKeysPhaseApplyLeader)
 	if err != nil {
 		return err
 	}
 	return generic.ErrSkip
 }
 
-// encryptionKeyRotationApplyLeader will run the next secrets-encrypt command, and scrape output to ensure that it was
-// successful. If the secrets-encrypt command does not exist on the plan, that means this is the first entrypoint and it
-// must be added, otherwise reenqueue until the plan is in sync.
-func (p *Planner) encryptionKeyRotationApplyLeader(cp *rkev1.RKEControlPlane, leader *planEntry) error {
-	var phase string
+// encryptionKeyRotationIsControlPlaneAndNotLeader allows us to filter cluster plans to restart healthy follower nodes
+func encryptionKeyRotationIsControlPlaneAndNotLeader(cp *rkev1.RKEControlPlane) roleFilter {
+	return func(entry *planEntry) bool {
+		return encryptionKeyRotationIsSuitableControlPlane(entry) &&
+			cp.Annotations[encryptionKeyRotationLeaderAnnotation] != entry.Machine.Name
+	}
+}
+
+func (p *Planner) encryptionKeyRotationRestartRKE2(cp *rkev1.RKEControlPlane, clusterPlan *plan.Plan, leader *planEntry) error {
+	if len(collect(clusterPlan, encryptionKeyRotationIsSuitableControlPlane)) == 1 {
+		return p.encryptionKeyRotationRestartSingleNodeAPIServer(cp, leader)
+	}
+	return p.encryptionKeyRotationRestartService(cp, clusterPlan, leader)
+}
 
-	_, err := encryptionKeyRotationExtractApplyFromPlan(leader)
+func (p *Planner) encryptionKeyRotationRestartK3S(cp *rkev1.RKEControlPlane, clusterPlan *plan.Plan, leader *planEntry) error {
+	return p.encryptionKeyRotationRestartService(cp, clusterPlan, leader)
+}
+
+func (p *Planner) encryptionKeyRotationRestartSingleNodeAPIServer(cp *rkev1.RKEControlPlane, leader *planEntry) error {
+	nodePlan := plan.NodePlan{
```

Comment:
If you can find a way to move this code into one helper function too, that'd be great

Reply:
This is the helper function, not sure it can be condensed any further